tests/test_registrable.py

```diff
@@ -1,27 +1,191 @@
-from pie_core.registrable import Registrable
+from typing import TypeVar
+
+import pytest
+
+from pie_core.registrable import Registrable, RegistrableProtocol, RegistrationError
+
+
+class A(Registrable):
+    """A base class."""
+
+    pass
+
+
+@A.register()
+class B(A):
+    """A registered subclass."""
+
+    pass
+
+
+class C(A):
+    """Unregistered subclass."""
+
+    pass
+
+
+class D(Registrable):
+    """Unregistered base class with no children."""
+
+    pass
 
 
 def test_registrable():
-    class A(Registrable):
+    """A basic Registrable class usage."""
+    # Check the registry entries are created
+    assert A in Registrable._registry
+    assert Registrable._registry[A] == {"B": B}
+
+    # Instantiate an Object of type B by its name.
+    clazz = A.by_name("B")()
+    assert isinstance(clazz, B)
+
+
+def test_base_class():
+
+    assert A.base_class() is A
+    assert B.base_class() is A
+    assert C.base_class() is A
+
+    with pytest.raises(RegistrationError) as e:
+        D.base_class()
+    assert str(e.value) == (
+        f"{D.__name__} has no defined base class. "
+        f"Please annotate {D.__name__} with @<SOME-PARENT-OF-{D.__name__}>.register() "
+        f"to register it at <SOME-PARENT-OF-{D.__name__}>."
+    )
+
+
+def test_has_base_class():
+    assert A.has_base_class()
+    assert B.has_base_class()
+    assert C.has_base_class()
+    assert not D.has_base_class()
+
+
+def test_register():
+    class Base(Registrable):
         pass
 
-    @A.register()
-    class B(A):
+    class Test(Base):
         pass
 
-    assert A in Registrable._registry
-    assert Registrable._registry[A] == {"B": B}
+    # Test main functionality
+    assert not Base.BASE_CLASS
+    assert not Test.BASE_CLASS
 
-    class_type = A.by_name("B")
-    assert class_type is B
+    wrapper = Base.register()
+    assert Base.BASE_CLASS
+    assert Test.BASE_CLASS
+    assert Base.base_class() is Base
+    assert Test.base_class() is Base
 
-    clazz = class_type()
-    assert isinstance(clazz, B)
+    # Test wrapper function
+    assert Registrable._registry[Base] == {}
+    wrapper(Test)
+    assert Registrable._registry[Base] == {"Test": Test}
 
-    assert A.registered_name_for_class(B) == "B"
-    assert A.registered_name_for_class(clazz.__class__) == "B"
 
-    class C(A):
+def test_register_subclass():
+    class Base(Registrable):
+        pass
+
+    @Base.register()
+    class Sub(Base):
         pass
 
+    with pytest.raises(RegistrationError) as e:
+        Sub.register()
+    assert str(e.value) == "Cannot register Sub; it is already registered as a subclass of Base"
```

Comment:
Why is this structured differently than below (here we call `Sub.register()`, there it uses `@Base.register()` above the class)?